Web Spec for Login with PayPal Button (AuthButton) (#55)
paypal/paypal-sdk-spec: spec/client/javascript/callbacks/onApprove-consent.md

# On Approve Consent

Callback used to signal partner once user accept the consent

- Successfully return auth_code and id_token

## Examples

### Get auth_code/id_token from paypal

```javascript
const onApprove = (data) => {
    return data; 
        // Data is an object which have two values either will be auth_code or id_token

>         // this return value depends on reponseType parameter 

Reviewer: There are quite a few typos in this doc, like reponseType should be responseType. Can you please take a look and fix these? This will be a public-facing document so needs to be exact :)

Author: done

# On Approve Consent

> Callback used to signal partner once user accept the consent

Reviewer 2: ```suggestion
Callback used to signal user consent approval from successfully completing the auth button flow.
```

Author: done

> - Successfully return auth_code and id_token based on responseType provided by partner.

Reviewer 2: I think we can drop this since it's explained by the examples below.
```suggestion
```

## Examples

> ### Get auth_code/id_token from paypal

Reviewer 2: Thinking we can keep this heading general to apply to both types of tokens.
```suggestion
### Get authorization tokens
```

Reviewer 2: I don't see this change made. @Karthickgreets99 can you take another look?

```javascript

> const onApprove = (data) => {

Reviewer 2: From my understanding, users of this auth button will send these tokens to their server-side api so they can later pull user info from the identity endpoints ex: https://developer.paypal.com/docs/api/identity/v1/#userinfo_get. If that's the case, can we illustrate it with a fetch call like so:
```js
const onApprove = (data, actions) => {
    // merchant sends auth tokens to their own api to
    // lookup additional user info using the paypal identity apis
    return fetch('/api/paypal/auth/consent', {
        method: 'POST',
        body: JSON.stringify({
            authCode: data.authCode
            idToken: data.idToken
        })
    }).then(res => {
        // Show a success message to the user
    });
};
```

Author: done

        // Data is an object which have two values either will be auth_code or id_token
        // this return value depends on reponseType parameter 
};
```
### Capture cancel consent code from paypal

```javascript
const onCancel = (data) => {
    // this value have the cancel code when the user cancel the consent.
};
```

## Types

```javascript
type OnApprove = (
    data : OnApproveData,
    actions : OnApproveActions
) => undefined | Promise<undefined>

type OnApproveData = {

>     authCode : string,

Reviewer: Will there always be an authCode even in idToken case? Or should this be `authCode? : string`?

Author: yes in idToken responseType we have authcode too in th object

# authorization tokens

Callback used to signal user consent approval from successfully completing the auth button flow.

## Examples

### Get authorization tokens from paypal

```javascript
const onApprove = (data, actions) => {
    // merchant sends auth tokens to their own api to
    // lookup additional user info using the paypal identity apis
    return fetch('/api/paypal/auth/consent', {
        method: 'POST',
        body: JSON.stringify({
            authCode: data.authCode,
            idToken: data.idToken
        })
    }).then(res => {
        // Show a success message to the user
    });
};
```
### Capture cancel consent code from paypal

```javascript
const onCancel = (data) => {
    // this value have the cancel code when the user cancel the consent.
};
```

## Types

```javascript
type OnApprove = (
    data : OnApproveData,
    actions : OnApproveActions
) => undefined | Promise<undefined>

type OnApproveData = {
    authCode : string,
    idToken? : string
};

> 

Reviewer 2: Can you add the types for the `data` payload for `onCancel`? I'm not sure the exact shape so it will be great to type it out. Something like:

```js
type OnCancelData = {
    errorCode: "",
    errorMessage: ""
};
```
